### PR TITLE
feat(authentication-service): add SAML passport authentication

### PR DESCRIPTION
**Pull request by RaghavaroraSF**
## Description
Added SAML passport authentication

Fixes # GH-1283

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] Performed a self-review of my own code
- [x] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: services/authentication-service/package.json [hunk 1]
```diff
@@ -23,7 +23,7 @@
     "eslint": "lb-eslint --report-unused-disable-directives .",
     "eslint:fix": "npm run eslint -- --fix",
     "openapi-spec": "node ./dist/openapi-spec",
-    "apidocs": "npx widdershins --language_tabs 'javascript:JavaScript:request' 'javascript--nodejs:Node.JS' --summary openapi.json -o openapi.md",
+    "apidocs": "./node_modules/.bin/widdershins --search false --language_tabs 'javascript:JavaScript:request' 'javascript--nodejs:Node.JS' --summary openapi.json -o openapi.md",
```
Comment (yeshamavani) [inline services/authentication-service/package.json:26]: ??

File: services/authentication-service/package.json [hunk 2]
```diff
@@ -62,18 +62,19 @@
...
     "bcrypt": "^5.0.1",
     "body-parser": "^1.20.0",
     "check-code-coverage": "^1.10.0",
     "cookie-parser": "^1.4.6",
-    "dotenv": "^16.0.3",
+    "dotenv": "^16.0.1",
     "dotenv-extended": "^2.9.0",
     "https-proxy-agent": "^5.0.1",
-    "jsonwebtoken": "^9.0.0",
+    "jsonwebtoken": "^8.5.1",
```
Comment (yeshamavani) [inline services/authentication-service/package.json:75]: why are we downgrading 
this is a vulnerable version.

File: services/authentication-service/package.json [hunk 3]
```diff
@@ -62,18 +62,19 @@
...
     "lodash": "^4.17.21",
-    "loopback4-authentication": "^7.4.1",
+    "loopback4-authentication": "file:loopback4-authentication-7.3.1.tgz",
```
Comment (yeshamavani) [inline services/authentication-service/package.json:77]: can not add this here.

File: services/authentication-service/package.json [hunk 4]
```diff
@@ -86,21 +87,21 @@
...
     "qrcode": "^1.5.0",
-    "tslib": "^2.4.1"
+    "tslib": "^2.4.0"
   },
   "devDependencies": {
     "@istanbuljs/nyc-config-typescript": "^1.0.2",
-    "@loopback/build": "^9.0.7",
-    "@loopback/eslint-config": "^13.0.7",
-    "@loopback/testlab": "^5.0.7",
+    "@loopback/build": "^9.0.3",
```
Comment (yeshamavani) [inline services/authentication-service/package.json:94]: again why are we downgrading these 
take the latest pull

File: services/authentication-service/package.json [hunk 5]
```diff
@@ -128,38 +129,12 @@
...
-    },
-    "tunnel-ssh": {
-      "debug": "^4.3.4"
-    },
-    "express": {
-      "debug": "^4.3.4",
-      "finalhandler": "^1.2.0",
-      "send": "^0.18.0",
-      "serve-static": "^1.15.0"
+    "passport-azure-ad": {
```
Comment (yeshamavani) [inline services/authentication-service/package.json:132]: no need of these overrides?

File: services/authentication-service/package.json [hunk 6]
```diff
@@ -128,38 +129,12 @@
...
-        "src/index.ts"
-      ],
-      "out": "services/authentication-service",
-      "plugin": [
-        "typedoc-plugin-markdown"
-      ]
-    },
-    "settings": {
-      "includeOpenAPIDoc": true
-    }
```
Comment (yeshamavani) [inline services/authentication-service/package.json:163]: ??

File: services/authentication-service/src/__tests__/unit/saml-post-verify.provider.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  describe('Post Verify Service', () => {
+    it('checks if provider returns a function', async () => {
+      const result = samlPostVerifyProvider.value();
+      expect(result).to.be.Function();
+    });
+
+    it('checks if provider function returns a promise which returns the user', async () => {
+      const func = samlPostVerifyProvider.value();
+      const result = await func(profile, user);
+      expect(result).to.be.eql(user);
```
Comment (yeshamavani) [inline services/authentication-service/src/__tests__/unit/saml-post-verify.provider.ts:39]: ??

File: services/authentication-service/src/modules/auth/providers/saml-verify.provider.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        } else {
+          throw new HttpErrors.Unauthorized(AuthErrorKeys.InvalidCredentials);
+        }
+      }
+      const creds = await this.userCredsRepository.findOne({
+        where: {
+          userId: user.id as string,
+        },
+      });
+      if (!creds || creds.authId !== profile.id) {
```
Comment (yeshamavani) [inline services/authentication-service/src/modules/auth/providers/saml-verify.provider.ts:55]: what are we saving as authProvider ?

File: services/authentication-service/src/modules/auth/saml-login.controller.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      authorizationURL: process.env.SAML_URL,
+      callbackURL: process.env.SAML_CALLBACK_URL,
+      clientID: process.env.SAML_CLIENT_ID,
+      clientSecret: process.env.SAML_CLIENT_SECRET,
+      tokenURL: process.env.SAML_TOKEN_URL,
+    },
+    queryGen('body'),
+  )
+  @authorize({permissions: ['*']})
+  @post('/auth/saml', {
```
Comment (yeshamavani) [inline services/authentication-service/src/modules/auth/saml-login.controller.ts:105]: get and post API endpoint should be same.

**Comment by yeshamavani** [comment]
lot of version downgrades in package.json file 
check those once

File: services/authentication-service/package.json (hunk 3 above)
Comment (RaghavaroraSF) [reply services/authentication-service/package.json:77]: resolved

File: services/authentication-service/package.json (hunk 2 above)
Comment (RaghavaroraSF) [reply services/authentication-service/package.json:75]: resolved

File: services/authentication-service/package.json (hunk 4 above)
Comment (RaghavaroraSF) [reply services/authentication-service/package.json:94]: resolved

File: services/authentication-service/package.json (hunk 5 above)
Comment (RaghavaroraSF) [reply services/authentication-service/package.json:132]: resolved

File: services/authentication-service/package.json (hunk 1 above)
Comment (RaghavaroraSF) [reply services/authentication-service/package.json:26]: resolved

File: services/authentication-service/package.json (hunk 6 above)
Comment (RaghavaroraSF) [reply services/authentication-service/package.json:163]: resolved

File: services/authentication-service/src/modules/auth/saml-login.controller.ts (hunk 9 above)
Comment (RaghavaroraSF) [reply services/authentication-service/src/modules/auth/saml-login.controller.ts:105]: resolved

File: services/authentication-service/src/modules/auth/providers/saml-verify.provider.ts (hunk 8 above)
Comment (RaghavaroraSF) [reply services/authentication-service/src/modules/auth/providers/saml-verify.provider.ts]: resolved

File: services/authentication-service/src/modules/auth/saml-login.controller.ts (hunk 9 above)
Comment (yeshamavani) [reply services/authentication-service/src/modules/auth/saml-login.controller.ts:66]: the deprecated @get  and @post should be same 
but the call-back one will be a different one 
refer others you will understand.

**Review by yeshamavani — CHANGES_REQUESTED** [review]

**Review by yeshamavani — APPROVED** [review]

**Comment by yeshamavani** [comment]
rebase the branch @RaghavaroraSF

**Pull request closed without merging**